### DISCUSSION TITLE
Проблемы с валидными значениями у DateInput

### DISCUSSION THREAD

**Discussion by IlyaAlexandrov1337** [VKCOM/VKUI#3505] [Q&A] [answered]
Нужно использовать DateInput как элемент выбора даты рождения, и возникло целых две проблемы с ограничением валидных дат: 

- если сверху их можно ограничить disableFuture, то снизу они должны ограничиваться определённым годом. Есть ли какая-то возможность это сделать нативно или придётся писать свой костыль?
-  сейчас можно вводить невалидные значения дня/месяца и проверка осуществляется уже после завершения всего ввода. Можно ли сделать так, чтобы пользователь не мог ввести с клавиатуры невалидные значения? (к примеру, сейчас можно ввести 33.33.3333 и пока не уберёшь фокус с DateInput, это значение так и будет висеть)

**Comment by inomdzhon** [comment, marked as answer]
**UPD**

1. Есть св-во `shouldDisableDate`. ~Изучил вопрос. Вижу, что возможности ограничить определённым годом, к сожалению, нет 😞  Создам задачу на добавление такой фичи. В данный момент я бы предложил ничего не костылить и дождаться фичи. Поговорить с дизайнером/продактом и объяснить, что смысла ограничивать пользователя на фронте нет. Ориентируйтесь на валидацию на бекенде, которая должна быть в любом случае.~
2. Текущее поведение меняться не будет. ~Можно будет ограничить, но не вижу проблемы в текущем решении – при расфокусе мы валидируем и сбрасываем если что-то не так. Через UI у пользователя нет возможности случайно вести не валидное значение. Опять же, валидация в любом случае должна быть на бекенде, ибо пользователь может через DevTools сделать всё что хочет, поэтому нельзя доверять валидацию клиенту.~

**Reply by IlyaAlexandrov1337** [reply]
2. Это тема UX: зачем пользователю давать возможность вводить некорректные данные, чтобы затем показывать ошибку/возвращать другую дату, если можно на этапе проектирования его ограничить в неправильном действии и сэкономить время всем. Дизайн топит именно за это.

**Reply by inomdzhon** [reply]
А зачем пользователю вводить неправильные данные? Разве что случайно, а если случайно, то пусть лучше увидит, что ввёл неправильно. Либо сам сразу поправит, либо мы.

Ты предлагаешь за пользователя это решать, т.к. это юзер френдли. Есть обратная сторона. Представь, ты пишешь «слепой печатью», нажимаешь цифру, а она не вводится. Нажимаешь повторно – она не вводится. Смотришь на клавиатуру, чтобы понять, что не так. Когда как ты сразу мог увидеть ошибку.

Сейчас работает как Т9, который может автоматически поправить твой текст, либо ты сам перепечатаешь.

**Reply by Zaycevq** [reply]
Давайте еще подключим @stoope, он занимался разработкой этого компонента.

Я соглашусь что ввод не валидных дат это неприятно, но кажется иначе сложнее осуществить, проверка после ввода одно из приемлемых решений.

**Reply by stoope** [reply]
1. Ограничить дату можно через коллбек `shouldDisableDate`
2. Очень сложно не давать пользователю вводить неправильные данные. Например, если юзер хочет ввести 31.10.2022 а у него стоит 01.02.2022 то в этом случае ему нельзя давать возможность ввести 31 число так как в феврале его нет? Ему сначала нужно будет сменить месяц а потом число?

**Reply by IlyaAlexandrov1337** [reply]
> 1. Ограничить дату можно через коллбек `shouldDisableDate`

Нет, нельзя. Вернее, `shouldDisableDate` ограничивает возможность выбрать дату в календаре, но по прежнему можно ввести с клавиатуры любую дату, `shouldDisableDate` будет просто игнорироваться в таком случае

**Reply by inomdzhon** [reply]
> > 1. Ограничить дату можно через коллбек `shouldDisableDate`
> 
> Нет, нельзя. Вернее, `shouldDisableDate` ограничивает возможность выбрать дату в календаре, но по прежнему можно ввести с клавиатуры любую дату, `shouldDisableDate` будет просто игнорироваться в таком случае

@Zaycevq @stoope а это баг или фича?
